lib/blocks.php

```diff
@@ -193,6 +196,105 @@ function gutenberg_register_core_block_styles( $block_name ) {
 	}
 }
 
+/**
+ * Change the way styles get loaded depending on their size.
+ *
+ * Optimizes performance and sustainability of styles by inlining smaller stylesheets.
+ *
+ * @return void
+ */
+function gutenberg_maybe_inline_styles() {
+
+	$total_inline_limit = 20000;
+	/**
+	 * The maximum size of inlined styles in bytes.
+	 *
+	 * @param int $total_inline_limit The file-size threshold, in bytes. Defaults to 20000.
+	 * @return int                    The file-size threshold, in bytes.
+	 */
+	$total_inline_limit = apply_filters( 'styles_inline_size_limit', $total_inline_limit );
+
+	global $wp_styles;
+	$styles = array();
+
+	// Build an array of styles that have a path defined.
+	foreach ( $wp_styles->queue as $handle ) {
+		if ( wp_styles()->get_data( $handle, 'path' ) && file_exists( $wp_styles->registered[ $handle ]->extra['path'] ) ) {
+			$block_styles = false;
+			$styles_size  = filesize( $wp_styles->registered[ $handle ]->extra['path'] );
+
+			// Minify styles and get their minified size if SCRIPT_DEBUG is not enabled.
+			if ( ! defined( 'SCRIPT_DEBUG' ) || ! SCRIPT_DEBUG ) {
+				// Get the styles and minify them by removing comments & whitespace.
+				$block_styles = gutenberg_get_minified_styles( file_get_contents( $wp_styles->registered[ $handle ]->extra['path'] ) );
+				// Get the styles size.
+				$styles_size = strlen( $block_styles );
+			}
+
+			$styles[] = array(
+				'handle' => $handle,
+				'path'   => $wp_styles->registered[ $handle ]->extra['path'],
+				'size'   => $styles_size,
+				'css'    => $block_styles,
+			);
+		}
+	}
+
+	if ( ! empty( $styles ) ) {
+		// Reorder styles array based on size.
+		usort(
+			$styles,
+			function( $a, $b ) {
+				return ( $a['size'] <= $b['size'] ) ? -1 : 1;
+			}
+		);
+
+		/**
+		 * The total inlined size.
+		 *
+		 * On each iteration of the loop, if a style gets added inline the value of this var increases
+		 * to reflect the total size of inlined styles.
+		 */
+		$total_inline_size = 0;
+
+		// Loop styles.
+		foreach ( $styles as $style ) {
+
+			// Size check. Since styles are ordered by size, we can break the loop.
+			if ( $total_inline_size + $style['size'] > $total_inline_limit ) {
+				break;
+			}
+
+			// Get the styles if we don't already have them.
+			$style['css'] = $style['css'] ? $style['css'] : file_get_contents( $style['path'] );
+
+			// Set `src` to `false` and add styles inline.
+			$wp_styles->registered[ $style['handle'] ]->src              = false;
+			$wp_styles->registered[ $style['handle'] ]->extra['after'][] = $style['css'];
+
+			// Add the styles size to the $total_inline_size var.
+			$total_inline_size += (int) $style['size'];
+		}
+	}
+}
+add_action( 'wp_head', 'gutenberg_maybe_inline_styles', 1 );
+
+/**
+ * Minify styles.
+ *
+ * Removes inline comments and whitespace.
+ *
+ * @param string $styles The styles to be minified.
+ * @return string
+ */
+function gutenberg_get_minified_styles( $styles ) {
```

Comment:
There's a bug in this function: You can easily see it by checking the "button" block in the "emptytheme". You'll see that the paddings are broken in the frontend.

While we could probably fix this somehow by improving the regex, I wonder if it's actually better to not minify ourselves and instead let the minification tools do the work by:

 1- Assuming that the file is already minified (and add it to our setup)
 2- Introduce a potential convention: load `.min.css` if the file is available instead of `.css`.

I personally lean towards  **1** and  potentially support for core blocks falling back in the blocks registration to unminified files if SCRIPT_DEBUG is true. (I don't like smart behaviors in general and we should better let the right tools handle the right job :) )

Reply:
I guess it's fixed by https://github.com/WordPress/gutenberg/pull/29554 that said I still think we'd better removing the minification from the framework itself.

Reply:
Regarding the bug, https://github.com/WordPress/gutenberg/pull/29554 was just merged and fixes it.

I agree however that we should provide minified files, the current ones are OK for debugging but on a real production environment, they are wasteful.

@youknowriad do you know _how_ to do that? I remember I tried to do it in https://github.com/WordPress/gutenberg/pull/25220 but ran into issues, mostly because I don't quite understand our tooling and how to use it 😅

Reply:
I think it's all in the `stylesTransform` function we have in the webpack config which in theory should generate minified styles for production code (`npm run build`) and unmagnified ones for development ( `npm run dev`)

Reply:
Huh... So it already runs as it should, I just always run dev so didn't see that they get minified on production builds 🤦 
I'll submit a PR to remove the minify function 👍

Reply:
At some point in the future, we might want to ship both files at the same time and introduce the `SCRIPT_DEBUG` flag support but it's not only about these blocks styles, it's about all JS and styles in Gutenberg plugin.

Reply:
> At some point in the future, we might want to ship both files at the same time and introduce the `SCRIPT_DEBUG` flag support but it's not only about these blocks styles, it's about all JS and styles in Gutenberg plugin.

This is how it works in the WordPress core.

Reply:
PR submitted on https://github.com/WordPress/gutenberg/pull/29624 👍

Reply:
Related issue about shipping `.min.*` files: https://github.com/WordPress/gutenberg/issues/23960